File: samples/demo-app/src/sampleModel.ts
import {configure} from '@sellwild/react-native-sdk';
import type {SellwildConfig} from '@sellwild/react-native-sdk';
import {eventQueue, setFailureContext} from '@sellwild/sdk-core';
import type {ClientFailureEvent, FailureSink} from '@sellwild/sdk-core';

// What the sample passes to the SDK. It is the same on every platform's
// sample app, so the e2e flows see the same app.
export const SampleSettings = {
  // Sellwild's own partner code. Never use a real partner's code here.
  partnerCode: 'sellwild',
  // There is no app config for this slug on the CDN (it answers 403). So
  // configure keeps the SDK's built-in config and reports config.fetch.http
  // once a launch. That is expected.
  slug: 'sellwild-sample',
  // Sellwild's own listings feed, passed as the public listingsUrl.
  listingsUrl: 'https://cache.sellwild.com/listings-img-data-sm-avif-fandom',
  // This app's bundle id (iOS) and application id (Android).
  appId: 'com.sellwild.sample.rn',
  // Zone ids for the ad slots when the config has none (the same as the
  // native samples).
  bannerZone: 'sellwild-sample-banner',
  mrecZone: 'sellwild-sample-mrec',
  // How long configure waits for the CDN.
  configTimeoutMs: 5000,
} as const;

/** Where the config came from: the CDN, or the SDK's built-in fallback. */
export type ConfigSource = 'remote' | 'fallback';

/** configure sets `remote` only when the CDN answered with a config. */
export function configSourceOf(config: SellwildConfig): ConfigSource {
  return config.remote ? 'remote' : 'fallback';
}

/**
 * The sample's values over what configure returned. The CDN wins for the
 * title and the zones when it has them; the configure overrides already set
 * the listings URL, the app id and debug.
 */
export function withSampleValues(config: SellwildConfig): SellwildConfig {
  return {
    ...config,
    title: config.title || 'Sellwild Sample',
    mobileZids:
      config.mobileZids.length > 0
        ? config.mobileZids
        : [SampleSettings.mrecZone],
    mobileBannerZid: config.mobileBannerZid || SampleSettings.bannerZone,
  };
}

/** What configure gave at launch. */
export interface SampleBoot {
  config: SellwildConfig;
  source: ConfigSource;
}

/** Runs configure once, with the sample's values. */
export async function bootSample(): Promise<SampleBoot> {
  const config = await configure(
    SampleSettings.partnerCode,
    SampleSettings.slug,
    {
      timeout: SampleSettings.configTimeoutMs,
      overrides: {
        listingsUrl: SampleSettings.listingsUrl,
        appBundleId: SampleSettings.appId,
        debug: true,
      },
    },
  );
  return {config: withSampleValues(config), source: configSourceOf(config)};
}

type CodesListener = (codes: readonly string[]) => void;

/**
 * The failure codes the SDK sent this launch, oldest first. Diagnostics
 * shows them. Listeners hear of a new code in a microtask: the SDK may report
 * while React renders.
 */
export class FailureLog {
  private codes: readonly string[] = [];
  private readonly listeners = new Set<CodesListener>();

  get current(): readonly string[] {
    return this.codes;
  }

  record(code: string): void {
    this.codes = [...this.codes, code];
    const codes = this.codes;
    Promise.resolve().then(() => {
      for (const listener of this.listeners) {
        listener(codes);
      }
    });
  }

  subscribe(listener: CodesListener): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }
}

/** Where a failure event goes after the log has seen it. */
export interface EventForwarder {
  push(event: ClientFailureEvent): void;
  flush(): void;
}

/**
 * The failure sink of the sample: it records each code, then sends the event
 * on to the SDK's events queue, as the SDK does when no sink is set. The SDK
 * calls it for each failure its gate lets through (after sampling and
 * dedupe), so the log holds what was sent.
 */
export function failureSink(
  log: FailureLog,
  forward: EventForwarder = eventQueue,
): FailureSink {
  return {
    push: event => {
      log.record(event.action);
      forward.push(event);
    },
    flushNow: () => forward.flush(),
  };
}

/** The app's one failure log. */
export const failureLog = new FailureLog();

/**
 * Makes failureLog the SDK's failure sink (setFailureContext, from
 * @sellwild/sdk-core). Call it before configure, so configure's own
 * failures are seen too. It sees what React Native JS and core report; the
 * native SDKs under the bridge report theirs without it.
 */
export function installFailureSink(): void {
  setFailureContext({sink: failureSink(failureLog)});
}

/**
 * The failure codes line: each code once, in the order first seen, with a
 * count when it was sent more than once.
 */
export function failureCodesText(codes: readonly string[]): string {
  if (codes.length === 0) {
    return 'none yet';
  }
  const counts = new Map<string, number>();
  for (const code of codes) {
    counts.set(code, (counts.get(code) ?? 0) + 1);
  }
  return [...counts]
    .map(([code, count]) => (count === 1 ? code : `${code} x${count}`))
    .join(', ');
}

/**
 * The listings status (contracts/e2e/ids.json, sw.listings.status). After a
 * good load: '<count> listings, load <n>', where n counts loads this launch.
 */
export function listingsStatus(
  count: number,
  loads: number,
  loading: boolean,
  error: Error | null,
): string {
  if (loading) {
    return `Loading listings (load ${loads + 1})`;
  }
  if (error) {
    return `Listings failed (load ${loads}): ${error.message}`;
  }
  return `${count} listings, load ${loads}`;
}

File: samples/flutter-demo/lib/src/sample_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:sellwild_sdk/sellwild_sdk.dart';

import 'sample_config.dart';

/// What configure gave at launch.
class SampleBoot {
  const SampleBoot(this.config, this.configSource);

  final SellwildConfig config;
  final ConfigSource configSource;
}

/// Runs SellwildSDK.configure once and keeps the failure codes the SDK sends
/// this launch (Diagnostics shows them).
class SampleModel {
  /// [send] passes each failure event on. The default sends it the way the
  /// SDK does without a sink: through SellwildAPIClient.instance.
  SampleModel({SellwildFailureSink? send}) : _send = send ?? sendThroughClient;

  final SellwildFailureSink _send;
  final List<String> _codes = [];

  /// The failure codes sent this launch, oldest first.
  final ValueNotifier<List<String>> failureCodes =
      ValueNotifier<List<String>>(const []);

  /// Installs the failure sink first, so configure's own failures are seen
  /// too, then runs configure with the sample's values.
  Future<SampleBoot> boot() async {
    installFailureSink();
    final config = await SellwildSDK.configure(
      partnerCode: SampleSettings.partnerCode,
      slug: SampleSettings.slug,
      overrides: withSampleValues,
    );
    return SampleBoot(config, configSourceOf(config));
  }

  /// Makes this model the SDK's failure sink (SellwildFailures.setContext).
  void installFailureSink() =>
      SellwildFailures.setContext(sink: _recordAndSend);

  /// The sink. The SDK calls it for each failure its gate lets through (after
  /// sampling and dedupe), so this is what reaches the events queue. The SDK
  /// can call it while Flutter builds a frame (a card reporting a bad photo),
  /// so listeners hear of a new code in a microtask, after the frame.
  Future<void> _recordAndSend(ClientFailureEvent event, bool flushNow) {
    _codes.add(event.action);
    scheduleMicrotask(() => failureCodes.value = List.unmodifiable(_codes));
    return _send(event, flushNow);
  }

  /// Sends [event] through SellwildAPIClient.instance, as the SDK does when
  /// no sink is set.
  static Future<void> sendThroughClient(ClientFailureEvent event, bool _) =>
      SellwildAPIClient.instance.sendEvent(
        event: event.event,
        action: event.action,
        label: event.label,
        uid: event.uid,
        attributes: event.attributes,
        createdTime: event.createdTime,
      );
}

/// The failure codes line: each code once, in the order first seen, with a
/// count when it was sent more than once.
String failureCodesText(List<String> codes) {
  if (codes.isEmpty) return 'none yet';
  final counts = <String, int>{};
  for (final code in codes) {
    counts[code] = (counts[code] ?? 0) + 1;
  }
  return counts.entries
      .map((e) => e.value == 1 ? e.key : '${e.key} x${e.value}')
      .join(', ');
}
